File: cn/TypeScript/121.Best Time to Buy and Sell Stock.ts
// @algorithm @lc id=121 lang=typescript
// @title best-time-to-buy-and-sell-stock
// @test([7,1,5,3,6,4])=5
// @test([7,6,4,3,1])=0
function maxProfit(prices: number[]): number {
  let minPrice = Infinity;
  let ans = 0;
  for (let p of prices) {
    ans = Math.max(ans, p - minPrice);
    if (p < minPrice) {
      minPrice = p;
    }
  }
  return ans;
}

File: cn/Python3/121_Best Time to Buy and Sell Stock.py
# @algorithm @lc id=121 lang=python3 
# @title best-time-to-buy-and-sell-stock


from cn.Python3.mod.preImport import *
# @test([7,1,5,3,6,4])=5
# @test([7,6,4,3,1])=0
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans, mi = 0, float('inf')
        for i, price in enumerate(prices):
            ans = max(ans, price - mi)
            mi = min(mi, price)
        return ans